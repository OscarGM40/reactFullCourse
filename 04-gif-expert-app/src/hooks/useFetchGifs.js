import { useEffect, useState } from "react";
import { getGifs } from "../helpers/getGifs";


//este hook almacena un objeto como estado.El objeto tiene dos props,un array llamado data y un boleano para el loading.El array lo llenara getGifs en base al valor del input(category).
export const useFetchGifs = (category) => {
  
   const [state, setState] = useState({
    data: [],
    loading: true,
  });


  useEffect(() => {
        
      setTimeout(() => {
           
           getGifs(category)
           .then(imgs => //imgs es el array que devuelve getGifs.
              setState({
                 data: imgs,
                 loading: false
              })
           );
        },200)

    }, [category]); 


  return state; //state es un objeto con un array y un boleano <- es el estado de este custom hook.El setter sólo lo usé aqui
};
