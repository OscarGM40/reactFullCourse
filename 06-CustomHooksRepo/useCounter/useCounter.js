import { useState } from "react";

const useCounter = (initialState = 10) => {
  const [counter, setCounter] = useState(initialState);

  const increment = (factor = 1) => {
    setCounter(counter + factor);
  };

  const decrement = (factor = 1) => {
    setCounter(counter - factor);
  };

  const reset = () => {
     setCounter(initialState);
  }

  //hay veces que es más fácil devolver un arreglo, otras veces será mejor devolver un objeto
  return {
    counter,
    increment,
    decrement,
    reset,
  };
};

export default useCounter;
