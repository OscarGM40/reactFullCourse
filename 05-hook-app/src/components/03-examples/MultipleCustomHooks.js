import React from "react";
import useCounter from "../../hooks/useCounter";
import { useFetch } from "../../hooks/useFetch";
import "../02-useEffect/effects.css";

const MultipleCustomHooks = () => {

  const { counter,increment,decrement } = useCounter(1);

  const { loading, data } = useFetch(`https://www.breakingbadapi.com/api/quotes/${ counter }`);
  //console.log(state)
  const { author ,quote } = !!data && data[0];
//console.log(data)
  

  

  return (
    <>
      <h1>Breaking Bad Quotes</h1>
      <hr />

      {
      loading ? 
         (
            <div className="alert alert-info text-center">Loading...</div>
         ) : 
         (
            <blockquote className="blockquote text-right mb-2">
               <p className="mb-0">{quote}</p>
               <footer className="blockquote-footer">{author}</footer>
            </blockquote>
         )
      }


      <button 
      className="btn btn-primary mb-0"
      onClick={() => increment(1)}>
         Siguiente Quote
      </button>

      <button 
      className="btn btn-primary mb-0"
      onClick={() => decrement(1)}>
   Anterior Quote
      </button>

    </>
  );
};

export default MultipleCustomHooks;
