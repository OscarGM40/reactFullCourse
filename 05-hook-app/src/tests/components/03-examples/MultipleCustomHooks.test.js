import { shallow } from "enzyme"
import MultipleCustomHooks from "../../../components/03-examples/MultipleCustomHooks"
import useCounter from "../../../hooks/useCounter";
import { useFetch } from "../../../hooks/useFetch";

jest.mock('../../../hooks/useFetch');
jest.mock("../../../hooks/useCounter");


describe('Pruebas en componente 03-examples/MultipleCustomHook', () => {
   
   beforeEach(() => {
   
      useCounter.mockReturnValue({
         counter:10,
         increment: () => {}
      })
   })

   test('Debe de hacer match con la snapshot', () => {
     
      useFetch.mockReturnValue({
         data:null,
         loading:true,
         error: null
      })

      const wrapper = shallow(<MultipleCustomHooks />)
      
      expect(wrapper).toMatchSnapshot();

   })

   test('Debe de mostrar la informacion', () => {
      
      useFetch.mockReturnValue({
         data:[{
            author:'Fernando',
            quote:'Hola mundo desde jest'
         }],
         loading:false,
         error: null
      })

      const wrapper = shallow(<MultipleCustomHooks />)

      expect(wrapper.find('.alert').exists()).toBeFalsy();
      expect(wrapper.find('p.mb-0').text().trim() ).toBe('Hola mundo desde jest');
      expect(wrapper.find('footer').text().trim() ).toBe('Fernando');

      console.log(wrapper.html())
   })
   
   
})
