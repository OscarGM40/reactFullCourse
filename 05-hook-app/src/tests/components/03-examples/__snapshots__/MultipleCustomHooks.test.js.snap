// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en componente 03-examples/MultipleCustomHook Debe de hacer match con la snapshot 1`] = `
<Fragment>
  <h1>
    Breaking Bad Quotes
  </h1>
  <hr />
  <div
    className="alert alert-info text-center"
  >
    Loading...
  </div>
  <button
    className="btn btn-primary mb-0"
    onClick={[Function]}
  >
    Siguiente Quote
  </button>
  <button
    className="btn btn-primary mb-0"
    onClick={[Function]}
  >
    Anterior Quote
  </button>
</Fragment>
`;
